### PR TITLE
feat(session-replay-browser): add page-level masking via urlMaskLevels (SR-3176)

### PR DESCRIPTION
**Pull request by lewgordon-amplitude**
## Summary

- Adds `urlMaskLevels` to `PrivacyConfig`, allowing masking level to be overridden per URL pattern (glob syntax)
- When the current page URL matches a `urlMaskLevels` rule, that rule's `maskLevel` is used as the baseline instead of `defaultMaskLevel`
- Per-element overrides (`amp-mask`, `maskSelector`, `unmaskSelector`) still apply on top, unchanged

## Changes

**`src/config/types.ts`** — Added `urlMaskLevels?: Array<{ match: string; maskLevel: MaskLevel }>` to `PrivacyConfig`.

**`src/helpers.ts`** — Added `getEffectiveMaskLevel(url, config)` using existing `globToRegex()` for pattern matching. Updated `isMasked`, `maskFn`, and `maskAttributeFn` to accept and forward the current URL for URL-aware masking decisions.

**`src/session-replay.ts`** — Tracks `currentPageUrl` (initialized on `init`, updated on SPA URL changes). Passes a URL getter to masking callbacks so they dynamically resolve the current URL. Updated `getMaskTextSelectors()` to use `getEffectiveMaskLevel`.

**`src/config/joined-config.ts`** — Concatenates `urlMaskLevels` during remote config merge (remote rules first for higher precedence in first-match order, then local).

**Tests** — 17 new tests covering `getEffectiveMaskLevel` (exact match, glob, first-match-wins, fallback, edge cases) and URL-aware `isMasked`/`maskFn` (per-element overrides still take priority, dynamic URL changes via getter).

## Test plan

- [x] All 658 existing + new tests pass
- [x] TypeScript compiles cleanly
- [ ] Manual verification with sample `urlMaskLevels` config in a test app

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes core masking behavior by making it URL-dependent and adds SPA URL tracking for masking/targeting, which could affect what data is captured if patterns are misconfigured or URL change detection fails.
> 
> **Overview**
> Adds **page-level masking overrides** via `PrivacyConfig.urlMaskLevels`, allowing the effective `maskLevel` to be selected by first-matching URL glob instead of always using `defaultMaskLevel`.
> 
> Updates masking helpers to compute `getEffectiveMaskLevel()` and thread the current URL through `isMasked`, `maskFn`, and `maskAttributeFn`, with `SessionReplay` tracking `currentPageUrl` across SPA navigations and passing a URL getter into rrweb’s mask callbacks. Remote/local config merging now concatenates `urlMaskLevels` with **remote rules first** (higher precedence), and tests are expanded to cover rule ordering, matching/fallback behavior, and dynamic URL changes.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit faac21795476412bc029ec51323d120b240d018a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Comment by lewgordon-amplitude** [comment]
bugbot run

File: packages/session-replay-browser/src/session-replay.ts [hunk 1]
```diff
@@ -172,6 +185,7 @@ export class SessionReplay implements AmplitudeSessionReplay {
     this.loggerProvider = new SafeLoggerProvider(options.loggerProvider || new Logger());
     Object.prototype.hasOwnProperty.call(options, 'logLevel') &&
       this.loggerProvider.enable(options.logLevel as LogLevel);
+    this.currentPageUrl = getCurrentUrl();
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/session-replay.ts:194]: Fixed in commit 02daadc3 on this branch: setupUrlChangeListener now runs unconditionally when urlMaskLevels has entries, not gated on targetingConfig existing. currentPageUrl is updated on every SPA navigation when URL masking rules are configured.

File: packages/session-replay-browser/src/session-replay.ts [hunk 2]
```diff
@@ -621,11 +635,14 @@ export class SessionReplay implements AmplitudeSessionReplay {
...
-    if (this.config?.privacyConfig?.defaultMaskLevel === 'conservative') {
+    const privacyConfig = this.config?.privacyConfig;
+    const effectiveLevel = privacyConfig ? getEffectiveMaskLevel(this.currentPageUrl, privacyConfig) : undefined;
+
+    if (effectiveLevel === 'conservative') {
       return '*';
     }
 
-    const maskSelector = this.config?.privacyConfig?.maskSelector;
+    const maskSelector = privacyConfig?.maskSelector;
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/session-replay.ts:660]: Documented known limitation — out of scope for this ticket. maskTextSelector is evaluated once by rrweb at recording start and cannot be updated mid-session without a recording restart. The conservative-any-rule-wins fallback in getMaskTextSelectors() handles the worst case (sessions that may visit a conservative page), with a minor perf tradeoff on non-conservative pages. A recording-restart approach is tracked as a follow-up.

File: packages/session-replay-browser/src/session-replay.ts [hunk 3]
```diff
@@ -621,11 +642,22 @@ export class SessionReplay implements AmplitudeSessionReplay {
...
+
+    if (effectiveLevel === 'conservative') {
+      return '*';
+    }
+
+    // If any urlMaskLevels rule uses 'conservative', always route all text nodes
+    // through maskTextFn so the dynamic URL getter can decide at call time.
+    // Without this, rrweb's static maskTextSelector would miss text nodes when
+    // the user navigates from a non-conservative page to a conservative one.
+    if (privacyConfig?.urlMaskLevels?.some((rule) => rule.maskLevel === 'conservative')) {
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/session-replay.ts:656]: Fixed in PR #1679 (sr-3176/url-mask-levels-additions), commit e7c67064: (1) isMaskedForLevel at 'light' now returns false for text nodes; (2) maskFn accepts a URL getter and calls getEffectiveMaskLevel() at call time, so masking decisions use the current page's effective level rather than defaultMaskLevel. The over-masking on light-URL pages is resolved.

File: packages/session-replay-browser/src/helpers.ts [hunk 4]
```diff
@@ -47,6 +47,21 @@ const isMaskedForLevel = (elementType: 'input' | 'text', level: MaskLevel, eleme
...
+export const getEffectiveMaskLevel = (url: string | undefined, config: PrivacyConfig): MaskLevel => {
+  if (url && config.urlMaskLevels) {
+    for (const rule of config.urlMaskLevels) {
+      if (globToRegex(rule.match).test(url)) {
+        return rule.maskLevel;
+      }
+    }
+  }
+  return config.defaultMaskLevel ?? DEFAULT_MASK_LEVEL;
+};
```
[the remark this answers is not in the document]
Comment (lewgordon-amplitude) [reply packages/session-replay-browser/src/helpers.ts:63]: Fixed in PR #1679 (sr-3176/url-mask-levels-additions), commit e7c67064: a module-level Map<string, RegExp> cache was added to globToRegex so each glob pattern is compiled to a RegExp once rather than on every maskTextFn/maskInputFn call.

**Comment by lewgordon-amplitude** [comment]
bugbot run

**Comment by lewgordon-amplitude** [comment]
Closing in favor of #1679, which supersedes this branch — it contains the same base feature plus the bug fixes (URL-aware maskFn, isMaskedForLevel light correction, High/Medium severity bugbot fixes) and full E2E coverage. The unique coverage commit here (faac2179) is already covered by e7c67064 on #1679.
